tensorflow_probability/python/distributions/stopping_ratio_logistic.py

```diff
@@ -0,0 +1,376 @@
+
+  dist.log_prob(1)
+  # ==> -0.82007515
+
+  dist.sample(3)
+  # ==> array([2, 1, 2], dtype=int32)
+  ```
+
+  """
+
+  def __init__(
+      self,
+      cutpoints,
+      loc,
+      dtype=tf.int32,
+      validate_args=False,
+      allow_nan_stats=True,
+      name='StoppingRatioLogistic',
+  ):
+    """Initialize Stopping Ratio Logistic distributions.
+
+    Args:
+      cutpoints: A floating-point `Tensor` with shape `(K,)` where
+        `K` is the number of cutpoints. The vector of cutpoints should be
+        non-decreasing, which is only checked if `validate_args=True`.
+      loc: A floating-point `Tensor` with shape `()`. The entry represents the
+        mean of the latent logistic distribution.
+      dtype: The type of the event samples (default: int32).
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g. mode) use the value "`NaN`" to indicate the result is
+        undefined. When `False`, an exception is raised if one or more of the
+        statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+    """
+
+    parameters = dict(locals())
+
+    with tf.name_scope(name) as name:
+
+      float_dtype = dtype_util.common_dtype(
+          [cutpoints, loc],
+          dtype_hint=tf.float32)
+
+      self._cutpoints = tensor_util.convert_nonref_to_tensor(
+          cutpoints, dtype_hint=float_dtype, name='cutpoints')
+      self._loc = tensor_util.convert_nonref_to_tensor(
+          loc, dtype_hint=float_dtype, name='loc')
+
+      super(StoppingRatioLogistic, self).__init__(
+          dtype=dtype,
+          reparameterization_type=reparameterization.NOT_REPARAMETERIZED,
+          validate_args=validate_args,
+          allow_nan_stats=allow_nan_stats,
+          parameters=parameters,
+          name=name)
+
+  @classmethod
+  def _params_event_ndims(cls):
+    return dict(cutpoints=1, loc=0)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return dict(
+        zip(('loc', 'scale'),
+            ([tf.convert_to_tensor(sample_shape, dtype=tf.int32)] * 2)))
+
+  @property
+  def cutpoints(self):
+    """Cutpoints parameter that separates the latent logistic distribution into
+    K + 1 categories."""
+    return self._cutpoints
+
+  @property
+  def loc(self):
+    """Mean parameter of the latent logistic distribution."""
+    return self._loc
+
+  def categorical_log_probs(self):
+    """Log probabilities for the `K + 1` sequential categories."""
+
+    cutpoints = self.cutpoints
+    loc = self.loc
+    num_cat = self._num_categories()
+
+    p = tf.math.log(tf.math.sigmoid(cutpoints - loc[..., tf.newaxis]))
+    q = tfm.log1mexp(p)
```

Comment:
Some of these expressions I would just put inline and rename the variables to be more descriptive (p and q are a little confusing, since we are also dealing with Categoricals, where q is normally 1 - p).